### PR TITLE
Feature/89 170 performance improvments

### PR DESCRIPTION
**Pull request by cschwinderg**
I was thinking about adding more performance improvements to this, but I decided that this was a good enough starting point. I want to see how long it takes to generate all of these indexes.

## Related Issues:
* [EQ-89](https://jira.epa.gov/browse/EQ-89)
* [EQ-170](https://jira.epa.gov/browse/EQ-170)

## Main Changes:
* Added indexes to any columns we plan to query against.
* Switched from using distinct queries to using recursive queries. This improves performance by emulating a loose scan index (currently not supported by postgres).
* Removed the `assessmentUnitStatus` field from tmdl, because this field is not in our tmdl table.
* Updated how our counts work. Apparently it is faster to do `count(*)` than it is to do `count(primary_key_column)`. 

## Steps To Test:
1. Add `CREATE_INDEXES_AFTER_DATA=true` to your `etl/.env.local` file
2. Run `npm run etl_database`
3. Verify the indexes were added
4. Verify the app runs

## TODO:
* [ ] Test the index creation in cloud.gov
* [ ] Get some index creation timing information from the logs
* [ ] Do some testing to see if there are indexes we can delete
* [ ] Further tune performance of commonly run queries (could involve creating materialized views)

### CODE REVIEW

File: etl/app/server/profiles/actions.js [hunk 1]
```diff
@@ -35,6 +36,73 @@ export const createQuery = `CREATE TABLE IF NOT EXISTS ${tableName}
...
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_assessmentunitid_asc
+    ON ${tableName} USING btree
+    (assessmentunitid COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_assessmentunitname_asc
+    ON ${tableName} USING btree
+    (assessmentunitname COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
```
Comment (maxdiebold-erg) [inline etl/app/server/profiles/actions.js:64]: The `completiondate` index is missing:
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_completiondate_desc
    ON ${tableName} USING btree
    (completiondate COLLATE pg_catalog."default" DESC NULLS LAST)
    TABLESPACE pg_default;
```

File: etl/app/server/profiles/assessments.js [hunk 2]
```diff
@@ -70,6 +71,218 @@ export const createQuery = `CREATE TABLE IF NOT EXISTS ${tableName}
...
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_associatedactionstatus_asc
+    ON ${tableName} USING btree
+    (associatedactionstatus COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_associatedactiontype_asc
+    ON ${tableName} USING btree
+    (associatedactiontype ASC NULLS LAST)
+    TABLESPACE pg_default;
+
```
Comment (maxdiebold-erg) [inline etl/app/server/profiles/assessments.js:139]: The `cyclefirstlisted` index is missing, not sure if ascending or descending is better in this case:
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_cyclefirstlisted_desc
    ON ${tableName} USING btree
    (cyclefirstlisted COLLATE pg_catalog."default" DESC NULLS LAST)
    TABLESPACE pg_default;
```

File: etl/app/server/profiles/assessments.js [hunk 3]
```diff
@@ -70,6 +71,218 @@ export const createQuery = `CREATE TABLE IF NOT EXISTS ${tableName}
...
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_cyclelastassessed_asc
```
Comment (maxdiebold-erg) [inline etl/app/server/profiles/assessments.js:140]: The `cycleexpectedtoattain` index is missing:
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_cycleexpectedtoattain_desc
    ON ${tableName} USING btree
    (cycleexpectedtoattain COLLATE pg_catalog."default" DESC NULLS LAST)
    TABLESPACE pg_default;
    
  CREATE INDEX IF NOT EXISTS ${indexTableName}_cyclelastassessed_asc
```

File: etl/app/server/profiles/assessments.js (hunk 2 above)
Comment (maxdiebold-erg) [inline etl/app/server/profiles/assessments.js:139]: The `consentdecreecycle` index is missing:
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_consentdecreecycle_desc
    ON ${tableName} USING btree
    (consentdecreecycle COLLATE pg_catalog."default" DESC NULLS LAST)
    TABLESPACE pg_default;
```

File: etl/app/server/profiles/catchmentCorrespondence.js [hunk 4]
```diff
@@ -25,6 +26,43 @@ export const createQuery = `CREATE TABLE IF NOT EXISTS ${tableName}
...
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_assessmentunitid_asc
+    ON ${tableName} USING btree
+    (assessmentunitid COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_assessmentunitname_asc
+    ON ${tableName} USING btree
+    (assessmentunitname COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
```
Comment (maxdiebold-erg) [inline etl/app/server/profiles/catchmentCorrespondence.js:39]: The `catchmentnhdplusid` index is missing. In the local database, these are all null, so I'm not sure how it reflects in the full dataset, or if we already discussed this during a meeting.
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_catchmentnhdplusid_asc
    ON ${tableName} USING btree
    (catchmentnhdplusid COLLATE pg_catalog."default" ASC NULLS LAST)
    TABLESPACE pg_default;
```

File: etl/app/server/profiles/tmdl.js [hunk 5]
```diff
@@ -46,6 +47,118 @@ export const createQuery = `CREATE TABLE IF NOT EXISTS ${tableName}
...
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_region_asc
+    ON ${tableName} USING btree
+    (region COLLATE pg_catalog."default" ASC NULLS LAST)
+    TABLESPACE pg_default;
+
+  CREATE INDEX IF NOT EXISTS ${indexTableName}_reportingcycle_desc
+    ON ${tableName} USING btree
+    (reportingcycle DESC NULLS LAST)
+    TABLESPACE pg_default;
+
```
Comment (maxdiebold-erg) [inline etl/app/server/profiles/tmdl.js:145]: The `state` index is missing:
```suggestion
CREATE INDEX IF NOT EXISTS ${indexTableName}_state_asc
    ON ${tableName} USING btree
    (state COLLATE pg_catalog."default" ASC NULLS LAST)
    TABLESPACE pg_default;
```

**Review by maxdiebold-erg — CHANGES_REQUESTED** [review]
Looks good, and the app still runs. Testing locally, I'm unable to see any performance differences.

There are a few missed/misspelled fields in the client-side configs, so I'll go ahead and push those changes.

I might suggest moving the `mapping` object out of `etl/app/server/database.js`, to separate data from logic.

## Other notes
- I noticed that we're indexing some of the columns with list values, like `delisted`, `inindiancountry`, `confirmed`, `includeinmeasure`, `pollutantindicator`, `usesupport`, and `vision303dpriority`, but not `cwa303dpriorityranking` or `organizationtype`. Was this a design choice?
- I also noticed that every index includes `NULLS LAST`, but I don't think this will cause a decrease in performance.
-  Since `reportingcycle` is required in our queries, we might benefit from creating composite indices with `reportingcycle` as the major column.

I noted some missing indices below.

**Comment by cschwinderg** [comment]
@maxdiebold-erg 

I skipped some of those indexes on purpose, because we don't have any UI elements for filtering on some of these fields. It is technically possible for people to still query against these fields by hitting the API directly. The idea for omitting some of these indexes is to save on disc space for indexes that are likely to never be used. Some of the indexes that I added may be overkill and we may need to remove some of these in the future. 

I see that the ones you mentioned above were missing because of configuration issues, so I'll go ahead and add in all of the indexes you mentioned above. 

Here are some responses to your questions/thoughts above
> I noticed that we're indexing some of the columns with list values, like `delisted`, `inindiancountry`, `confirmed`, `includeinmeasure`, `pollutantindicator`, `usesupport`, and `vision303dpriority`, but not `cwa303dpriorityranking` or `organizationtype`. Was this a design choice?
This was done on purpose, to save on disc space for fields that are likely to never be queried against. 

> I also noticed that every index includes `NULLS LAST`, but I don't think this will cause a decrease in performance.
Postgres gives you an option to either do `NULLS LAST` or `NULLS FIRST`. In our case we are never going to be querying for nulls, so we should get a little performance boost by putting the nulls at the end. 

> Since `reportingcycle` is required in our queries, we might benefit from creating composite indices with `reportingcycle` as the major column.
I considered doing composite indexes, but the extra disc space isn't really worth it. Postgres can utilize multiple indexes in a single query plan. I didn't see much of a performance gain when compairing composite indexes vs having postgres use multiple indexes.

Should the TMDL profile have the "State" filter? I noticed it isn't there.

**Comment by maxdiebold-erg** [comment]
@cschwinderg That makes sense. For the `NULLS LAST` point, I realize the benefit, I just meant it's being included on non-nullable fields, but still shouldn't hurt.

It looks like "State" is in the TMDL table, I'll go ahead and add that to the filters.

**Review by maxdiebold-erg — APPROVED** [review]

**Pull request merged**